node_modules/col-activities/lib/notifications/weekly.js

```diff
@@ -0,0 +1,583 @@
+      } else {
+        // There is no recipient separate from the actor. Count points generated only.
+        userTotals.incrementPoints('generated', points);
+      }
+    }
+  });
+
+  // 2. Divide course totals by the user count to get averages.
+  summary.course.averages = _.mapValues(summary.course.totals, function(courseTotal) {
+    return _.round(courseTotal / userCount);
+  });
+
+  // 3. Find the top asset in each category.
+  summary.course.topAssets = {};
+  _.forEach(['comments', 'likes', 'views'], function(key) {
+    var topAsset = sampleMaximum(summary.assets, _.property('weeklyTotals.' + key));
+    if (topAsset) {
+      summary.course.topAssets[key] = summary.assets[topAsset.id];
+    }
+  });
+
+  // 4. Find the top user in each category.
+  summary.course.topUsers = {};
+  _.forEach(['pointsGenerated', 'pointsReceived'], function(key) {
+    var topUser = sampleMaximum(summary.users, _.property(key));
+    if (topUser) {
+      summary.course.topUsers[key] = {
+        'total': topUser.value,
+        'user': users[topUser.id]
+      };
+    }
+  });
+
+  // TODO: Iterate through assets to get keyword frequencies from titles and descriptions (if created this
+  // week), and from comments.
+
+  return summary;
+};
+
+/**
+ * Return or instantiate running totals for an asset.
+ * 
+ * @param  {Number}    asset      The asset to get totals for
+ * @param  {Object}    summary    The activities summary
+ * @return {AssetTotals}          Running totals for the given asset
+ * @api private
+ *
+ */
+var getAssetTotals = function(asset, summary) {
+  if (!asset) {
+    return null;
+  }
+  if (!_.has(summary.assets, asset.id)) {
+    // Assets, unlike users, need their full property set stored in the summary as the email template will
+    // make use of them. Add asset properties to the summary, including a new AssetTotals property for the
+    // week.
+    asset = asset.toJSON();
+    asset.weeklyTotals = new AssetTotals();
+    summary.assets[asset.id] = asset;
+
+    // Each asset added to the summary also needs to be accessible from per-user totals, for purposes of.
+    // calculating most popular asset per user.
+    _.each(asset.users, function(assetUser) {
+      var assetUserTotals = getUserTotals(assetUser.id, summary);
+      assetUserTotals.assets[asset.id] = asset;
+    });
+  }
+
+  return summary.assets[asset.id].weeklyTotals;
+};
+
+/**
+ * Return or instantiate running totals for a user.
+ * 
+ * @param  {Number}    userId     The user id to get totals for
+ * @param  {Object}    summary    The activities summary
+ * @return {UserTotals}           Running totals for the given user
+ * @api private
+ *
+ */
+var getUserTotals = function(userId, summary) {
+  if (!_.has(summary.users, userId)) {
+    summary.users[userId] = new UserTotals();
+  }
+  return summary.users[userId];
+};
+
+/**
+ * Running totals for activites and points per asset.
+ */
+var AssetTotals = function() {
```

Comment:
Thanks for the pointer, @nicolaasmatthijs. Just to be clear, you're endorsing the "functional pattern" over the "pseudoclassical pattern" despite the drawbacks to the functional pattern listed in the linked article? I was particularly wary of expensively recreating functions on every object instantiation, though perhaps this is less of a problem with modern interpreters.